proposals/0088-enable-core-bpf-programs.md

```diff
@@ -0,0 +1,113 @@
+---
+simd: '0088'
+title: Enable Core BPF Programs
+authors:
+  - Joe Caulfield
+category: Standard
+type: Core
+status: Draft
+created: 2023-11-07
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+This proposal introduces the concept of Core BPF programs: programs which are
+essential to network operations. Currently, these exist as built-in programs
+known as "native" programs.
+
+This SIMD details the process by which existing native programs can be
+ported to Core BPF, as well as the process for introducing brand-new Core BPF
+programs.
+
+## Motivation
+
+BPF programs offer less complexity than native programs for other clients, such
+as Firedancer, since developers will no longer have to keep up with program
+changes in their runtime implementations. Instead, the program can just be
+updated once.
+
+For this reason, it makes sense to introduce the concept of Core BPF programs:
+BPF programs the network depends on that should be treated with special care.
+
+## Alternatives Considered
+
+The alternative to Core BPF programs is to keep these essential programs as
+native programs. This would mean each validator client implementation would have
+to build and maintain these built-in programs with their runtime
+implementations, including any future changes to these programs introduced via
+SIMDs or other fixes.
+
+## New Terminology
+
+- `Core BPF Program`: A BPF program relied on by any part of the Solana stack,
+  including (but not limited to) consensus, transaction processing, voting,
+  staking, and account creation.
+
+## Detailed Design
+
+Core BPF programs in many ways will be designed no differently than any other
+BPF program on Solana. However, some programs may require special privileges,
+which is beyond the scope of this SIMD.
+
+When an existing native program is being proposed to be migrated to Core BPF,
+or when a new Core BPF program is being introduced, at least one SIMD shall be
+published outlining at least the following details:
+
+- How any required special privileges will be granted to the program in its BPF
+  form
+- How this program's upgrades will be managed after it becomes Core BPF
+
+**Migrating a native program to core BPF** shall consist of deploying a
+BPF version of the native program to a new arbitrary address and using a
+feature gate to move the BPF program, replacing the existing native program at
+its program address.
+
+In the context of this design, **target program** refers to an existing native
+program, while **source program** refers to the BPF version to be moved into the
+existing program account.
+
+The migration process must adhere to the following steps:
+
+1. Verifiably build the source BPF program.
+2. Generate a new keypair for the source program.
+3. Deploy the program to the source program address.
```

Comment:
Can't we just use the same address?

Reply:
I think we need to replace the program at the original address anyway, otherwise all its owned accounts will not be available to the replacement.

Reply:
To replace a native program with a BPF program, you'd have to deploy the BPF version to some arbitrary address, then move it into the account at the original address. That would result in the address being unchanged and the program being replaced with BPF.

See [my comment here](https://github.com/solana-foundation/solana-improvement-documents/pull/88#discussion_r1441929536).

Reply:
That is fine.